### ISSUE TITLE
Phase createMergedModule fails with module not found on an explicit dependency

### ISSUE THREAD

**Issue by doriancransac** [beryx/badass-jlink-plugin#172]
I'm having issues with the org.slf4j dependency ever since I added an intermediate module in my module hierarchy.

```
:pv-gameserver:createMergedModule
  Cannot derive uses clause from service loader invocation in: com/fasterxml/jackson/databind/ObjectMapper.secureGetServiceLoader().
  Cannot derive uses clause from service loader invocation in: com/fasterxml/jackson/databind/ObjectMapper$2.run().
  Cannot derive uses clause from service loader invocation in: org/junit/jupiter/engine/extension/ExtensionRegistry.registerAutoDetectedExtensions().
  /mnt/agent/work/ac708666ab59f0c/pv-gameserver/build/jlinkbase/tmpjars/com.pv.gameserver.merged.module/module-info.java:409: error: module not found: org.slf4j
      requires org.slf4j;
                         ^
  1 error
```

I'm not entirely sure where the jar came from before or at least why I wasn't getting any errors but I do have an explicit dependency to slf4j in my root build.gradle file (for all projects) which should apply:

```
        implementation "org.slf4j:slf4j-api:$2.0.0-alpha1"
        runtime "org.slf4j:slf4j-simple:$2.0.0-alpha1"
```
with the corresponding requires in the module-info.java:
```
requires org.slf4j;
```

I tried everything I could find online in other similar issues and in the documentation, i.e:
```
    forceMerge("com.pv.the_intermediate_module")
    addExtraDependencies("org.slf4j")
```
I also tried `forceMerge("org.slf4j")`, which didn't work either.

In some cases the plugin completes but then I get different errors at runtime similar to [this](https://github.com/beryx/badass-jlink-plugin/issues/127). To solve that one, I ended up using this:

```
    mergedModule {
        additive = true
        requires("org.slf4j")
    }
```

**BUT, NONE** of these directives result in the slf4j api jar being actually included in the merge. The only workaround I was able to find is by adding a completely arbitrary third party dependency which I'm using in a _different_ (and completely unrelated) module, after noticing that this module did not suffer from the problem, while still using the same libs resulting in calls made to slf4j. When this dependency is active, then the prepareMergedJars phase pulls the jar into the jlinkjars folder and then everything works, even at runtime but needless to say this is a very ugly and random workaround as this module doesn't actually need this dependency.

```
compile("org.glassfish.jersey.connectors:jersey-jetty-connector:3.0.0")
```
with the corresponding requires in the module-info.java:
```
requires jersey.jetty.connector;
```

Any idea what could be wrong here?

**Comment by siordache** [comment]
The arguments of the `forceMerge` method are prefixes of the names of the JAR files to be merged. In your case, these are: slf4j-api-2.0.0-alpha1.jar and slf4j-simple-2.0.0-alpha1.jar, so you should use `forceMerge("slf4j")` instead of `forceMerge("org.slf4j")`.
Alternatively, you can avoid calling `forceMerge("slf4j")` by using non-modular releases of the slf4j libraries:
```
implementation 'org.slf4j:slf4j-api:1.7.30'
runtime 'org.slf4j:slf4j-simple:1.7.30'
```

**Comment by doriancransac** [comment]
@siordache 

Ah thanks, that makes a lot of sense. I think I initially went with the 1.8-beta (and then later the 2.0 due to subsequent issues) version because of the classic split package syndrome. Am I understanding correctly that my new problem here (and in particular the IllegalAccessError at runtime caused by my mongodb driver attempting to access the LoggerFactory) is related to the fact that I was using modular slf4j BUT that the unnamed module tries to access it?

I think I'm still getting kind of confused between the different nuances of unnamed module, automatic module and merged module (i.e how the plugin ends up making a decision of putting what jar where).

So am I understanding correctly that in this transitional phase (which I'm starting to think may just last forever), not only do we have to worry about split packages but even when a library tries to do the "right" thing (like slf4j with versions >= 1.8-beta), we can still be held back because of some OTHER non-modular libs referencing them?

Thanks again for taking the time to post this answer, I appreciate it very much. I'll give the different options a shot right away.

**Comment by siordache** [comment]
Yes, that's correct. The problem was the unnamed module trying to access the modular slf4j.
The plugin [tries to detect such cases](https://badass-jlink-plugin.beryx.org/releases/latest/#_how_it_works) and to automatically include the problematic modular jars into the merged module. But sometimes it fails to detect that a jar should be treated as non-modular. In this case, you need to use `forceMerge`.

When running gradle with the `-i` option, the `prepareMergedJarsDir` task shows which dependencies are modular and which are non-modular (or handled as non-modular). This might help to troubleshoot the build.

```
modularJars: [...]
nonModularJars: [...]
modularJarsRequiredByNonModularJars: [...]
artifactsHandledAsNonModular: [...]
```

**Comment by doriancransac** [comment]
I just removed the hacky dependency and applied `forceMerge("slf4j")` as recommended and it works. On a side note, I HAD to remove this:

```
mergedModule {
        additive = true
        requires("org.slf4j")
    }
```

Otherwise I'd still be getting the module not found error (not entirely sure why, but at least it works). I do remember that I fixed the original module not found by migrating from 1.8-beta to 2.0-beta but I can't remember the exact circumstances of it appearing in the first place (again, I've been working with 1.8-beta for a few months, but only recently, as I started to add more modules did this problem show up again, not sure why).

Btw, It might be a good idea to mention explicitly in the documentation that the argument of forceMerge is a jar file name prefix, not a package prefix (I know jars are mentioned in the paragraphs but not that the prefix matches the file name, which confused me personally).
